api/pkg/transformer/feast/bigtablestore/encoding.go

```diff
@@ -0,0 +1,462 @@
+				status[fr] = serving.GetOnlineFeaturesResponse_NOT_FOUND
+			}
+			fieldValues[i] = &serving.GetOnlineFeaturesResponse_FieldValues{
+				Fields:   fields,
+				Statuses: status,
+			}
+			continue
+		}
+		for _, fr := range req.Features {
+			featureRef, err := ParseFeatureRef(fr)
+			if err != nil {
+				return nil, err
+			}
+			maxAge := e.featureMetadata[featureTableKey{
+				project: req.Project,
+				table:   featureRef.FeatureTable,
+			}].MaxAge
+			if maxAge != nil && maxAge.GetSeconds() > 0 && timestamp[featureTableKey{
+				project: req.Project,
+				table:   featureRef.FeatureTable,
+			}].Add(time.Duration(maxAge.GetSeconds())*time.Second).Before(time.Now()) {
+				fields[fr] = &types.Value{}
+				status[fr] = serving.GetOnlineFeaturesResponse_OUTSIDE_MAX_AGE
+				fieldValues[i] = &serving.GetOnlineFeaturesResponse_FieldValues{
+					Fields:   fields,
+					Statuses: status,
+				}
+				continue
+			}
+
+			avroValue := avroValueByKey[bigtableKey][featureTableKey{
+				project: req.Project,
+				table:   featureRef.FeatureTable,
+			}][featureRef.Feature]
+
+			featureSpec := e.featureSpecs[featureTableKey{
+				project: req.Project,
+				table:   featureRef.FeatureTable,
+			}]
+
+			featureType, err := getFeatureType(e.featureTypes, featureSpec.Project, fr)
+			if err != nil {
+				return nil, err
+			}
+			val, err := avroToValueConversion(avroValue, featureType)
+			if err != nil {
+				return nil, err
+			}
+			fields[fr] = val
+			status[fr] = serving.GetOnlineFeaturesResponse_PRESENT
+		}
+		fieldValues[i] = &serving.GetOnlineFeaturesResponse_FieldValues{
+			Fields:   fields,
+			Statuses: status,
+		}
+	}
+
+	return &feast.OnlineFeaturesResponse{
+		RawResponse: &serving.GetOnlineFeaturesResponse{
+			FieldValues: fieldValues,
+		},
+	}, nil
+}
+
+// CodecRegistry returns an Avro codec given a schema reference hash in byte array
+type CodecRegistry interface {
+	GetCodec(ctx context.Context, schemaRef []byte, project string, entityKeys []*spec.Entity) (*goavro.Codec, error)
+}
+
+// CachedCodecRegistry instantiates an Avro codec by retrieving the schema from
+// BigTable. The Avro schema corresponds to a FeatureTable is always stored as
+// a row within the same BigTable, where the key is the schema reference hash. The
+// codec is cached to avoid repeated calls to BigTable.
+type CachedCodecRegistry struct {
+	codecs map[string]*goavro.Codec
+	tables map[string]storage
+	sync.RWMutex
+}
+
+// newCachedCodecRegistry instantiates a new CachedCodecRegistry
+func newCachedCodecRegistry(tables map[string]storage) *CachedCodecRegistry {
+	return &CachedCodecRegistry{
+		codecs: make(map[string]*goavro.Codec),
+		tables: tables,
+	}
+}
+
+// GetCodec returns an Avro codec, either from the in-memory cache or built
+// from the Avro schema specification.
+func (r *CachedCodecRegistry) GetCodec(ctx context.Context, schemaRef []byte, project string, entityKeys []*spec.Entity) (*goavro.Codec, error) {
+	r.RLock()
```

Comment:
Why do you use RLock/RUnlock here but Lock/Unlock below?

Reply:
using `RLock` enable multiple go routine to read (not write), that's why we use it on the read part. For write part we only enable one goroutine by using `Lock`

Reply:
TIL